### DISCUSSION TITLE
Frustrating migration experience

### DISCUSSION THREAD

**Discussion by AlleSchonWeg** [dotnet/maui#19900] [Upgrading from Xamarin]
After three weeks of "migration" from XF (Android/iOS) to Maui (.Net8) my frustration level is becoming higher. Almost nothing works. 
I spend two weeks to make my app compiling again. Commented out some code/nugets which are not available for maui or using workarounds:

- ObservableRangeCollection (Toolkit) --> ToDo 
- BadgeView (Toolkit) --> Workaround https://github.com/CommunityToolkit/Maui/issues/105
- LocalizedString (Toolkit) --> Workaround  https://github.com/SirJohnK/LocalizationResourceManager.Maui
- TouchEffect (Toolkit) --> ToDo https://github.com/CommunityToolkit/Maui/issues/86
- LifecycleEffect (Toolkit) --> ToDo
- Push: (https://github.com/CrossGeeks/FirebasePushNotificationPlugin) --> https://github.com/thomasgalliker/Plugin.FirebasePushNotifications

Compiling VS (17.8.4) Errors. Sometimes build again (multiple times) or restart VS works or delete bin/obj works or both:

- ArrayMapKt.class error --> https://github.com/dotnet/maui/issues/14954
- Microsoft.AppCenter.Crashes clang++ exited with code 1  --> https://github.com/microsoft/appcenter-sdk-dotnet/issues/1755
- Long Path problem --> https://github.com/xamarin/GoogleApisForiOSComponents/issues/639 (multiple issues opened everywhere who uses this component)
- IDE1100 Error reading content of source file 'C:\maui\src\XXX\obj\Debug\net8.0-android\XXX.AssemblyInfo.cs' -- 'Could not find file 'C:\maui\src\XXX\obj\Debug\net8.0-android\XXX.AssemblyInfo.cs'.'
- MAUI0000	System.IO.IOException: The process cannot access the file 'C:\maui\src\XXX\obj\Debug\net8.0-android\resizetizer\r\drawable-xhdpi\YYYY.png' because it is being used by another process.
- Error	XAGR7028	System.IO.FileNotFoundException: Could not find file 'C:\maui\src\XXX\obj\Debug\net8.0-android\lp\218\jl\res\values-tl\values.xml'.
- VS randomly decided to change the .csproj file. --> https://github.com/dotnet/maui/issues/17993

If i have luck and the project compiles, app starts and the UI appears. My first impressions:
- Systemfonts not working --> https://github.com/dotnet/maui/issues/19556
- Switch looks different --> https://github.com/dotnet/maui/issues/19676
- Toolbaritem can't be removed --> https://github.com/dotnet/maui/issues/3701
- MonkeyCache crashes --> https://github.com/jamesmontemagno/monkey-cache/issues/136
- Android Button Ripple Effect broken --> https://github.com/dotnet/maui/issues/17642
- Clipping broken --> https://github.com/dotnet/maui/issues/15138
- Rotating device crashes the app -->` java.lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.` I don't know why
- ListView is jittering while scrolling --> I don't know why
- Navigation between pages is not smooth --> I don't know why

This happens on the "first" pages in my app. On other pages the UI is "destroyed". 

To be fair, not all is bad:
- MinimumHeightRequest / MinimumWidthRequest works as expected.
- Android App Icon (Adaptive Icon etc.) generation 
- Displaying SVG's works very nice. 

Edit:
- Cant access BottomNavigationView in TabbedViewHandler to modify LabelVisibilityMode. In XF it works https://montemagno.com/xamarin-forms-fully-customize-bottom-tabs-on-android-turn-off-shifting/

**Comment by jkommeren** [comment]
Well wait until you discover firebase notifications don't work on Android in release mode. (I saw a firebase thing in your long path problem, which I saw too)
Unless you turn off both AOT and trimming.

Not sure why it's not listed as a known issue when it's so hard to figure out why notifications are not arriving.

It took me a week.

**Reply by Xyncgas** [reply]
yikes

**Reply by maonaoda** [reply]
> Not sure why it's not listed as a known issue when it's so hard to figure out why notifications are not arriving.

they don't use it themselves

**Reply by Ghevi** [reply]
I also confirm this

**Comment by AlleSchonWeg** [comment]
> Well wait until you discover firebase notifications don't work on Android in release mode. (I saw a firebase thing in your long path problem, which I saw too) Unless you turn off both AOT and trimming.
> 
> Not sure why it's not listed as a known issue when it's so hard to figure out why notifications are not arriving.
> 
> It took me a week.

Thank you for the warning. Do you have a link to StackOverflow or to an issue which describes the behavior?

**Comment by jkommeren** [comment]
Any time, I would really love this issue to be fixed!

Here's the ticket I could find on MAUI, though I'm not sure it covers the exact problem: https://github.com/dotnet/maui/issues/16014 
I'm not sure it's related to AOT or Trimming myself, not sure how this guy is sure it's due to the Trimming.

Here's a SO ticket, with a suggestion of how to set up trimming with linker. (Note that with the suggested linking doesn't seem to work for me, at least I'm not confident it DOES work, but the mechanism is working)
https://stackoverflow.com/questions/76045900/net-maui-why-would-push-notifications-not-work-the-same-after-publishing

It's hard to narrow down, because either MAUI is framework is working against me (bin/obj stuff remaining after rebuild), or Google is simply not sending notifications any more due to too many registration attempts or something when installing/uninstalling the app? Also, each build takes minutes due to AOT and Trimming.

Last thing I've done is I managed to get the dependencies to the latest version (partly looking at what ThomasGalliker has in his source code), that doesn't solve the problem, my hope was with https://github.com/thomasgalliker/Plugin.FirebasePushNotifications you mentioned as well, but I think he's going to run into the same issue once he's done with his refactoring PR. If it does work for you please let me know.

Easy way to test: Set to release mode and deploy on device (trimming and AOT are on by default), also I'm assuming TrimOnPublish also applies to Release thing, not specifically Publishing it (as I can see it's performing trimming during release build)

**Comment by Xyncgas** [comment]
Why not use WebUI, generate PDF in .NET and display iframe of pdf in browser makes so much sense than doing it in xamarin
